# First look at the data

In this notebook I study the dataset and prepare it for the analysis.
There is almost no business logics here, but this step is necessary for further EDA.

In [1]:
import pandas as pd

In [2]:
# Reading the data
dict_data = pd.read_excel("online_retail_II.xlsx", sheet_name = None)
# We have several sheets, so let's keep them in one dataset
data = pd.concat(dict_data.values())
data = data.reset_index(drop=True)
# Loading from .xlsx is very slow, let's store data in .hdf.
# data.to_hdf("sales.hdf",key="df")
# And read from hdf in future versions:
# data = pd.read_hdf("sales.hdf")

### Describing the dataset

In [3]:
# Checking the number of entries
data.shape

(1067371, 8)

In [4]:
# First look at the data
data

Invoice StockCode                          Description  Quantity  \
0        489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1        489434    79323P                   PINK CHERRY LIGHTS        12   
2        489434    79323W                  WHITE CHERRY LIGHTS        12   
3        489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4        489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
...         ...       ...                                  ...       ...   
1067366  581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
1067367  581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
1067368  581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
1067369  581587     22138        BAKING SET 9 PIECE RETROSPOT          3   
1067370  581587      POST                              POSTAGE         1   

                InvoiceDate  Price  Customer ID         Country  
0       2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1       2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2       2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3       2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4       2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
...                     ...    ...          ...             ...  
1067366 2011-12-09 12:50:00   2.10      12680.0          France  
1067367 2011-12-09 12:50:00   4.15      12680.0          France  
1067368 2011-12-09 12:50:00   4.15      12680.0          France  
1067369 2011-12-09 12:50:00   4.95      12680.0          France  
1067370 2011-12-09 12:50:00  18.00      12680.0          France  

[1067371 rows x 8 columns]

In [5]:
# Describing dataset
data.head(10)

Invoice StockCode                          Description  Quantity  \
0  489434     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS        12   
1  489434    79323P                   PINK CHERRY LIGHTS        12   
2  489434    79323W                  WHITE CHERRY LIGHTS        12   
3  489434     22041         RECORD FRAME 7" SINGLE SIZE         48   
4  489434     21232       STRAWBERRY CERAMIC TRINKET BOX        24   
5  489434     22064           PINK DOUGHNUT TRINKET POT         24   
6  489434     21871                  SAVE THE PLANET MUG        24   
7  489434     21523   FANCY FONT HOME SWEET HOME DOORMAT        10   
8  489435     22350                            CAT BOWL         12   
9  489435     22349       DOG BOWL , CHASING BALL DESIGN        12   

          InvoiceDate  Price  Customer ID         Country  
0 2009-12-01 07:45:00   6.95      13085.0  United Kingdom  
1 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
2 2009-12-01 07:45:00   6.75      13085.0  United Kingdom  
3 2009-12-01 07:45:00   2.10      13085.0  United Kingdom  
4 2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
5 2009-12-01 07:45:00   1.65      13085.0  United Kingdom  
6 2009-12-01 07:45:00   1.25      13085.0  United Kingdom  
7 2009-12-01 07:45:00   5.95      13085.0  United Kingdom  
8 2009-12-01 07:46:00   2.55      13085.0  United Kingdom  
9 2009-12-01 07:46:00   3.75      13085.0  United Kingdom

In [6]:
# Describing numeric columns
data.describe(datetime_is_numeric=True)

Quantity                    InvoiceDate         Price  \
count  1.067371e+06                        1067371  1.067371e+06   
mean   9.938898e+00  2011-01-02 21:13:55.394028544  4.649388e+00   
min   -8.099500e+04            2009-12-01 07:45:00 -5.359436e+04   
25%    1.000000e+00            2010-07-09 09:46:00  1.250000e+00   
50%    3.000000e+00            2010-12-07 15:28:00  2.100000e+00   
75%    1.000000e+01            2011-07-22 10:23:00  4.150000e+00   
max    8.099500e+04            2011-12-09 12:50:00  3.897000e+04   
std    1.727058e+02                            NaN  1.235531e+02   

         Customer ID  
count  824364.000000  
mean    15324.638504  
min     12346.000000  
25%     13975.000000  
50%     15255.000000  
75%     16797.000000  
max     18287.000000  
std      1697.464450

In [7]:
# Describing "Object" columns
data[['Invoice','StockCode','Description','Country']].describe()

Invoice StockCode                         Description         Country
count   1067371   1067371                             1062989         1067371
unique    53628      5305                                5698              43
top      537434    85123A  WHITE HANGING HEART T-LIGHT HOLDER  United Kingdom
freq       1350      5829                                5918          981330

### Deeper look

In [8]:
# Let's add Amount = Quantity * Price column right away
# In this way we will get an idea of order of money sent to the company - just curiosity.
data['Amount'] = data['Price']*data['Quantity']
data['Amount'].sum()

19287250.56799999

In [9]:
# Now, let's look at anomalies
data.query("Quantity < -80000 or Quantity > 80000")

Invoice StockCode                  Description  Quantity  \
1065882   581483     23843  PAPER CRAFT , LITTLE BIRDIE     80995   
1065883  C581484     23843  PAPER CRAFT , LITTLE BIRDIE    -80995   

                InvoiceDate  Price  Customer ID         Country    Amount  
1065882 2011-12-09 09:15:00   2.08      16446.0  United Kingdom  168469.6  
1065883 2011-12-09 09:27:00   2.08      16446.0  United Kingdom -168469.6

In [10]:
# Unique values 
data.nunique()

Invoice        53628
StockCode       5305
Description     5698
Quantity        1057
InvoiceDate    47635
Price           2807
Customer ID     5942
Country           43
Amount          9113
dtype: int64

In [11]:
# Check the description
data['Description'].value_counts()

WHITE HANGING HEART T-LIGHT HOLDER    5918
REGENCY CAKESTAND 3 TIER              4412
JUMBO BAG RED RETROSPOT               3469
ASSORTED COLOUR BIRD ORNAMENT         2958
PARTY BUNTING                         2765
                                      ... 
found box                                1
HEXAGONAL CRYSTAL PERFUME BOTTLE         1
sold as set by dotcom                    1
SET OF 4 KNICK KNACK TINS  DOILEY        1
thrown away-can't sell                   1
Name: Description, Length: 5698, dtype: int64

Some descriptions are apparently hand-written. Good to know

In [12]:
# Let's look at all countries
a = data['Country'].value_counts()
for _ in a.index:
    print(_)

United Kingdom
EIRE
Germany
France
Netherlands
Spain
Switzerland
Belgium
Portugal
Australia
Channel Islands
Italy
Norway
Sweden
Cyprus
Finland
Austria
Denmark
Unspecified
Greece
Japan
Poland
USA
United Arab Emirates
Israel
Hong Kong
Singapore
Malta
Iceland
Canada
Lithuania
RSA
Bahrain
Brazil
Thailand
Korea
European Community
Lebanon
West Indies
Bermuda
Nigeria
Czech Republic
Saudi Arabia


Some countries are missing. Again, good to know.

In [13]:
# We know that invoice numbers are either integer or start from "C" - for cancellation.
# Is it really always the case?
a = data['Invoice'].value_counts()
def test_number(i):
    global data
    try:
        int(i)
        return
    except:
        pass
    if i[0] == "C":
        try:
            int(i[1:])
            return
        except:
            pass
    print(i)
    return
        
for invoice_number in a.index:
    test_number(invoice_number)
        

A506401
A516228
A563186
A563185
A528059
A563187


No, there are some weird "A"-starting invoices. Let's take a closer look.

In [14]:
data.query("Invoice == 'A563187'")

Invoice StockCode      Description  Quantity         InvoiceDate  \
825445  A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

           Price  Customer ID         Country    Amount  
825445 -11062.06          NaN  United Kingdom -11062.06

This is adjustment of bad debts. Apparently, that's standard business practice and they are put in base with negatove price. Let's look at other entries with negative price.

In [15]:
data.query("Price < 0")

Invoice StockCode      Description  Quantity         InvoiceDate  \
179403  A506401         B  Adjust bad debt         1 2010-04-29 13:36:00   
276274  A516228         B  Adjust bad debt         1 2010-07-19 11:24:00   
403472  A528059         B  Adjust bad debt         1 2010-10-20 12:04:00   
825444  A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
825445  A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

           Price  Customer ID         Country    Amount  
179403 -53594.36          NaN  United Kingdom -53594.36  
276274 -44031.79          NaN  United Kingdom -44031.79  
403472 -38925.87          NaN  United Kingdom -38925.87  
825444 -11062.06          NaN  United Kingdom -11062.06  
825445 -11062.06          NaN  United Kingdom -11062.06

### Conclusion

Few things we see so far:
 * We have over one million entries (1067371)
 * We have around 50k invoices shipped (53628)
 * We have almost 6k customers (5942)
 * The customers have spent almost 20 millon pounds (19287250)
 * Customers are from 43 countries
 * Data needs cleaning:
   * Negative prices caused by debth adjusting
   * Hand-written and missing product descriptions, different product descriptions for the same stock code
   * Wrong countries ("Unspecified", "European Community")
   * Entries with non-existing customers IDs
   
Before we proceed any further, we need to clean the data. We aim to perform customer-centric analysis, so we may ignore for now wrong or missing product descriptions. What we will do is the following:

 * Remove all entries without customer ID. Though this is significant part of the dataset (almost 20%), we simply can't use it for customer-centric analysis. This will also eliminate "Adjust bad debt" entries
 * Account for cancellations: for each cancelled entry find the most recent buy of the customer of the product with the same `StockCode` and reduce the quantity by the quantity of the cancellation. Recalculate `Amount` once done.

## Preparing data

### Removing entries without customer ID

In [16]:
# I don't like column names with spaces inside.
data['cID'] = data['Customer ID']

In [17]:
# This is simple way to check non-nan values
data = data.query("cID == cID")

### Dealing with cancellations

Description of the dataset explicitly states that there are "cancellations" entries. I don't want to use them in customer analysis (though it may be interesting to know when cancellations occure, and, may be, take some actions to prevent them - by improving design of the web-interfae, for example). 

Cancellation items have invoice number starting from "C". Moreover I assume that cancellations have the same user name and occures after the initial order (this is the meaning of "cancellation"). The cancellations can be complete or partial. Some of the cancellations may relate to sales outside of the dataset (i.e. to those happened before the first entry of the database).

For each cancellation, I want to find an original invoice that should be done earlier by the same customer and contain the same item. Once original invoice is found, I'll reduce the quantities accordingly. If no original invoice is found, I ignore the cancellation. I remove all cancellation entries from the dataset in the end.

In [18]:
# Invoice number of cancelled shipping is not relevant, so we can drop it
# Afte this line Invoice number will be integer, and cancellations will have not-a-number value.
data['Invoice'] = pd.to_numeric(data['Invoice'], errors = 'coerce')

/Users/ilya/Env_For_All/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [19]:
# This dataframe of cancellations
cancellations = data.query("Invoice != Invoice")

In [20]:
# This dataframe doesn't have cancellations
corrected_data = data.query("Invoice == Invoice")

In [21]:
corrected_data['old_index'] = corrected_data.index

/Users/ilya/Env_For_All/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


In [22]:
# I merge both datasets by cliend ID and stock code. This dataset contains all possible permutations.
auxdata =  pd.merge(corrected_data, cancellations[['Quantity','cID','StockCode','InvoiceDate']],\
                    on=['cID','StockCode'], suffixes=('', '_c'))

In [23]:
# The cancelled amount can't be higher than ordered
auxdata = auxdata.query('Quantity >= -1*Quantity_c')

In [24]:
# Cancellaction comes after the order
auxdata = auxdata.query('InvoiceDate<=InvoiceDate_c')

In [25]:
# In case we have several orders with the same item from the same user made before cancellation, 
# we pick the most recent one.
auxdata['Time_to_correction'] = (auxdata['InvoiceDate_c'] - auxdata['InvoiceDate']).astype('timedelta64[D]')
auxdata=auxdata.loc[auxdata.groupby("Invoice")["Time_to_correction"].idxmin()]

In [26]:
# Some of the cancellations are made quite late. Let's assume that cancellation should be done within one year
auxdata = auxdata.query("Time_to_correction<365")

In [27]:
# Now, let's add the cancellations to corrected data
corrected_data = corrected_data.join(auxdata[['Quantity_c','old_index']].set_index('old_index'), on = 'old_index')
corrected_data['Quantity_c'] = corrected_data['Quantity_c'].fillna(0)

In [28]:
# Get the corrected quantities
corrected_data['Quantity'] = corrected_data['Quantity_c'] + corrected_data['Quantity']

In [29]:
# And recalculate amount
corrected_data['Amount'] = corrected_data['Quantity']*corrected_data['Price']

In [30]:
corrected_data.describe()

Invoice       Quantity          Price    Customer ID  \
count  805620.000000  805620.000000  805620.000000  805620.000000   
mean   537410.885564      12.939181       3.206279   15331.856250   
std     26666.396589      75.099906      29.197901    1696.768395   
min    489434.000000       0.000000       0.000000   12346.000000   
25%    514962.000000       2.000000       1.250000   13982.000000   
50%    536989.000000       5.000000       1.950000   15271.000000   
75%    561618.000000      12.000000       3.750000   16805.000000   
max    581587.000000   19152.000000   10953.500000   18287.000000   

              Amount            cID     old_index     Quantity_c  
count  805620.000000  805620.000000  8.056200e+05  805620.000000  
mean       21.179297   15331.856250  5.314914e+05      -0.368484  
std        67.435445    1696.768395  3.083349e+05     123.210505  
min         0.000000   12346.000000  0.000000e+00  -80995.000000  
25%         4.890000   13982.000000  2.644758e+05       0.000000  
50%        11.700000   15271.000000  5.161975e+05       0.000000  
75%        19.500000   16805.000000  8.075002e+05       0.000000  
max      8925.000000   18287.000000  1.067370e+06       0.000000

In [31]:
corrected_data

Invoice StockCode                          Description  Quantity  \
0        489434.0     85048  15CM CHRISTMAS GLASS BALL 20 LIGHTS      12.0   
1        489434.0    79323P                   PINK CHERRY LIGHTS      12.0   
2        489434.0    79323W                  WHITE CHERRY LIGHTS      12.0   
3        489434.0     22041         RECORD FRAME 7" SINGLE SIZE       48.0   
4        489434.0     21232       STRAWBERRY CERAMIC TRINKET BOX      24.0   
...           ...       ...                                  ...       ...   
1067366  581587.0     22899         CHILDREN'S APRON DOLLY GIRL        6.0   
1067367  581587.0     23254        CHILDRENS CUTLERY DOLLY GIRL        4.0   
1067368  581587.0     23255      CHILDRENS CUTLERY CIRCUS PARADE       4.0   
1067369  581587.0     22138        BAKING SET 9 PIECE RETROSPOT        3.0   
1067370  581587.0      POST                              POSTAGE       1.0   

                InvoiceDate  Price  Customer ID         Country  Amount  \
0       2009-12-01 07:45:00   6.95      13085.0  United Kingdom   83.40   
1       2009-12-01 07:45:00   6.75      13085.0  United Kingdom   81.00   
2       2009-12-01 07:45:00   6.75      13085.0  United Kingdom   81.00   
3       2009-12-01 07:45:00   2.10      13085.0  United Kingdom  100.80   
4       2009-12-01 07:45:00   1.25      13085.0  United Kingdom   30.00   
...                     ...    ...          ...             ...     ...   
1067366 2011-12-09 12:50:00   2.10      12680.0          France   12.60   
1067367 2011-12-09 12:50:00   4.15      12680.0          France   16.60   
1067368 2011-12-09 12:50:00   4.15      12680.0          France   16.60   
1067369 2011-12-09 12:50:00   4.95      12680.0          France   14.85   
1067370 2011-12-09 12:50:00  18.00      12680.0          France   18.00   

             cID  old_index  Quantity_c  
0        13085.0          0         0.0  
1        13085.0          1         0.0  
2        13085.0          2         0.0  
3        13085.0          3         0.0  
4        13085.0          4         0.0  
...          ...        ...         ...  
1067366  12680.0    1067366         0.0  
1067367  12680.0    1067367         0.0  
1067368  12680.0    1067368         0.0  
1067369  12680.0    1067369         0.0  
1067370  12680.0    1067370         0.0  

[805620 rows x 12 columns]

In [32]:
corrected_data.to_hdf("cleaned_sales.hdf",key="df")

/Users/ilya/Env_For_All/lib/python3.7/site-packages/pandas/core/generic.py:2446: PerformanceWarning: 
your performance may suffer as PyTables will pickle object types that it cannot
map directly to c-types [inferred_type->mixed-integer,key->block3_values] [items->Index(['StockCode', 'Description', 'Country'], dtype='object')]

  encoding=encoding,
